### PR TITLE
Migrating focusable tooltip component in V9

### PR DESCRIPTION
**Pull request by srmukher**
Migrating focusable tooltip component in V9

### CODE REVIEW

File: packages/react-components/react-charts-preview/src/utilities/FocusableTooltipText.tsx [hunk 1]
```diff
@@ -1,80 +1,66 @@
+/* eslint-disable no-restricted-globals */
```
Comment (AtishayMsft) [inline packages/react-components/react-charts-preview/src/utilities/FocusableTooltipText.tsx:1]: why added this rule?

File: packages/react-components/react-charts-preview/src/utilities/KeyCodes.ts [hunk 2]
```diff
@@ -0,0 +1,111 @@
...
+  capslock: 20 as 20,
+  escape: 27 as 27,
+  space: 32 as 32,
+  pageUp: 33 as 33,
+  pageDown: 34 as 34,
+  end: 35 as 35,
+  home: 36 as 36,
+  left: 37 as 37,
+  up: 38 as 38,
+  right: 39 as 39,
```
Comment (AtishayMsft) [inline packages/react-components/react-charts-preview/src/utilities/KeyCodes.ts:24]: are we using all of these keys

File: packages/react-components/react-charts-preview/src/utilities/ChartHoverCard/ChartHoverCard.styles.ts [hunk 3]
```diff
@@ -10,7 +10,7 @@ export const getChartHoverCardStyles = (props: IChartHoverCardStyleProps): IChar
         display: 'grid',
         overflow: 'hidden',
         padding: '11px 16px 10px 16px',
-        backgroundColor: theme.semanticColors.bodyBackground,
+        backgroundColor: tokens.colorNeutralBackground1,
```
Comment (AtishayMsft) [inline packages/react-components/react-charts-preview/src/utilities/ChartHoverCard/ChartHoverCard.styles.ts:13]: are these changes sufficient to get chart hover card to work?

File: packages/react-components/react-charts-preview/src/utilities/FocusableTooltipText.tsx [hunk 4]
```diff
@@ -1,80 +1,66 @@
...
 
-  private _getTargetElement = (): HTMLElement | undefined => {
-    if (!this._tooltipChild.current || !this._tooltipChild.current.parentElement) {
-      return undefined;
-    }
+    return () => {
+      resizeObserver.current?.disconnect();
+      async.dispose();
+    };
+  }, [async, checkTextOverflow, getTargetElement]);
```
Comment (AtishayMsft) [inline packages/react-components/react-charts-preview/src/utilities/FocusableTooltipText.tsx:55]: why is async function a dependency of the effect hook

File: packages/react-components/react-charts-preview/src/utilities/FocusableTooltipText.tsx [hunk 5]
```diff
@@ -1,80 +1,66 @@
...
-      this._resizeObserver = new ResizeObserver(this._async.debounce(this._checkTextOverflow, 500));
-      this._resizeObserver.observe(overflowElement);
+  const checkTextOverflow = useCallback(() => {
+    const overflowElement = getTargetElement();
+    const isTextOverflow = !!overflowElement && hasOverflow(overflowElement);
+    if (isTextOverflow !== textOverflow) {
+      setTextOverflow(isTextOverflow);
     }
-  }
+  }, [getTargetElement, textOverflow, setTextOverflow]);
```
Comment (AtishayMsft) [inline packages/react-components/react-charts-preview/src/utilities/FocusableTooltipText.tsx:38]: why is getTargetElement a dependency of this callback

File: packages/react-components/react-charts-preview/src/utilities/FocusableTooltipText.tsx [hunk 6]
```diff
@@ -1,80 +1,66 @@
...
+      resizeObserver.current.observe(overflowElement);
     }
-  };
 
-  private _getTargetElement = (): HTMLElement | undefined => {
-    if (!this._tooltipChild.current || !this._tooltipChild.current.parentElement) {
-      return undefined;
-    }
+    return () => {
+      resizeObserver.current?.disconnect();
```
Comment (AtishayMsft) [inline packages/react-components/react-charts-preview/src/utilities/FocusableTooltipText.tsx:52]: how are you mimicking unmount behavior in this effect

File: packages/react-components/react-charts-preview/src/utilities/FocusableTooltipText.tsx [hunk 7]
```diff
@@ -1,80 +1,66 @@
...
+    };
+  }, [async, checkTextOverflow, getTargetElement]);
 
-    return this._tooltipChild.current.parentElement;
-  };
-}
+  return (
+    <div className={props.className}>
+      <Tooltip content={props.content} relationship="label">
+        <span {...getAccessibleDataObject(props.accessibilityData)} ref={tooltipChild} data-is-focusable={textOverflow}>
```
Comment (AtishayMsft) [inline packages/react-components/react-charts-preview/src/utilities/FocusableTooltipText.tsx:60]: is this block accessible?

File: packages/react-components/react-charts-preview/src/utilities/ChartHoverCard/ChartHoverCard.styles.ts (hunk 3 above)
Comment (srmukher) [reply packages/react-components/react-charts-preview/src/utilities/ChartHoverCard/ChartHoverCard.styles.ts:13]: yes, hover card is working. Will just merge it with popover component in next PR

File: packages/react-components/react-charts-preview/src/utilities/FocusableTooltipText.tsx (hunk 1 above)
Comment (srmukher) [reply packages/react-components/react-charts-preview/src/utilities/FocusableTooltipText.tsx:1]: without this rule it says "window" cannot be used. we have to import it from useFluent() but that was also throwing error

File: packages/react-components/react-charts-preview/src/utilities/FocusableTooltipText.tsx (hunk 5 above)
Comment (srmukher) [reply packages/react-components/react-charts-preview/src/utilities/FocusableTooltipText.tsx:38]: overflowElement requires getTargetElement()

File: packages/react-components/react-charts-preview/src/utilities/KeyCodes.ts (hunk 2 above)
Comment (srmukher) [reply packages/react-components/react-charts-preview/src/utilities/KeyCodes.ts:24]: no o we are using only Escape, but kept all of them for reference

File: packages/react-components/react-charts-preview/src/utilities/ChartHoverCard/ChartHoverCard.styles.ts (hunk 3 above)
Comment (AtishayMsft) [reply packages/react-components/react-charts-preview/src/utilities/ChartHoverCard/ChartHoverCard.styles.ts:13]: you are using v8 dependencies in chart hover card. They need to be removed.

File: packages/react-components/react-charts-preview/src/utilities/FocusableTooltipText.tsx (hunk 4 above)
Comment (srmukher) [reply packages/react-components/react-charts-preview/src/utilities/FocusableTooltipText.tsx:55]: becuase of the async.dispose() Other wise it gives error "React Hook useEffect has a missing dependency: 'async'. Either include it or remove the dependency array."

File: packages/react-components/react-charts-preview/src/utilities/FocusableTooltipText.tsx (hunk 6 above)
Comment (srmukher) [reply packages/react-components/react-charts-preview/src/utilities/FocusableTooltipText.tsx:52]: The useEffect hook does not execute the entire effect on every lifecycle phase. Instead, it runs the effect on mount and update, and the cleanup function on unmount.
On Mount:
  - Call getTargetElement()
  - If ResizeObserver is available and overflowElement is not null:
    - Create new ResizeObserver with debounced checkTextOverflow
    - Observe overflowElement

On Update (when async, checkTextOverflow, or getTargetElement change):
  - Call getTargetElement()
  - If ResizeObserver is available and overflowElement is not null:
    - Create new ResizeObserver with debounced checkTextOverflow
    - Observe overflowElement

On Unmount:
  - Disconnect ResizeObserver
  - Dispose async

File: packages/react-components/react-charts-preview/src/utilities/KeyCodes.ts (hunk 2 above)
Comment (srmukher) [reply packages/react-components/react-charts-preview/src/utilities/KeyCodes.ts:24]: done

File: packages/react-components/react-charts-preview/src/utilities/ChartHoverCard/ChartHoverCard.styles.ts (hunk 3 above)
Comment (srmukher) [reply packages/react-components/react-charts-preview/src/utilities/ChartHoverCard/ChartHoverCard.styles.ts:13]: yes that is left, haven't started with chart hover card yet.

Comment (AtishayMsft) [reply packages/react-components/react-charts-preview/src/utilities/ChartHoverCard/ChartHoverCard.styles.ts:13]: This is the only error that is left during build now

Comment (srmukher) [reply packages/react-components/react-charts-preview/src/utilities/ChartHoverCard/ChartHoverCard.styles.ts:13]: ok will start next

File: packages/react-components/react-charts-preview/src/utilities/FocusableTooltipText.tsx (hunk 6 above)
Comment (srmukher) [reply packages/react-components/react-charts-preview/src/utilities/FocusableTooltipText.tsx:52]: Verified as follows: 
Setup/update: Setup block is called when the chart load. 
Unmount: Used a button to unmount the chart. when the button is clicked the cleanup block gets executed.

File: packages/react-components/react-charts-preview/src/utilities/FocusableTooltipText.tsx (hunk 7 above)
Comment (srmukher) [reply packages/react-components/react-charts-preview/src/utilities/FocusableTooltipText.tsx:61]: the tooltip itself is not accessible but the text is accessible via keyboard.

File: packages/react-components/react-charts-preview/src/utilities/FocusableTooltipText.tsx (hunk 6 above)
Comment (AtishayMsft) [reply packages/react-components/react-charts-preview/src/utilities/FocusableTooltipText.tsx:53]: I am not getting where is the logic to restricted to run on unmount. I get that the dispose logic is executed when resize observer disconnects. Lets discuss this more. Could you log a bug for this.
Meanwhile approving this PR to unblock the build.

**Review by AtishayMsft — APPROVED** [review]

File: packages/react-components/react-charts-preview/src/utilities/FocusableTooltipText.tsx (hunk 6 above)
Comment (srmukher) [reply packages/react-components/react-charts-preview/src/utilities/FocusableTooltipText.tsx:53]: This article explains in detail https://react.dev/reference/react/useEffect#reference regarding how the setup and cleanup parts work within the useeffect.

**Pull request merged**